File: Solution in CPP/Que7.cpp
// Implement a function that returns the kth smallest element in a given array.
// assuming k starts from 1 and k is always less than array size.
#include<bits/stdc++.h>
using namespace std;

int finder(vector<int> v, int k){
    map<int, int> mp;
    for(auto it: v){
        mp[it]++;
    }

    for(auto it: mp){
        while(it.second != 0){
            it.second --;
            k --;
            if(k == 0){
                return it.first;
            }
        }
    }

    return v[0];
}

int main(){
    vector<int> array = {7,5,3,1,1,2,6,8,11,2,6,7};
    int k = 2; // find the second smallest element

    // sorting the vector.
    int ans = finder(array, k);
    cout<<k<<"th smallest element in the given array is: "<<ans<<endl;



}

File: Solution in RUST/Que7.rs
// Implement a function that returns the kth smallest element in a given array.


use std::collections::HashMap;

fn finder(v: Vec<i32>, k: usize) -> i32 {
    let mut mp: HashMap<i32, usize> = HashMap::new();
    for &num in &v {
        *mp.entry(num).or_insert(0) += 1;
    }

    let mut current_k = 0;
    for i in 0..=10000 {
        if let Some(&count) = mp.get(&(i as i32)) {
            current_k += count;
            if current_k >= k {
                return i as i32;
            }
        }
    }

    panic!("Invalid input: k exceeds the number of unique elements");
}

fn main() {
    let array = vec![7, 5, 3, 1, 1, 2, 6, 8, 11, 2, 6, 7];
    let k = 12; // find the second smallest element

    let ans = finder(array, k);
    println!("{}th smallest element in the given array is: {}", k, ans);
}
